### PR TITLE
[NoQA] Fix the internal QA PR section losing its state

### PR DESCRIPTION
**Pull request by mountiny**
### Details

The internal QA section of the Staging deploy [issue](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3AStagingDeployCash) loses its state when the issue is regenerated when new deploy blocker is found or we CP some PR to Staging.

Not ideal.

This PR aims to fix this situation by passing a list of URLs of the InternalQA PRs which were checked off already to the function which generates the new issue body.

I have left comments in the PR to explain how this is fixing the problem and also why the integration test hasn't caught this issue previously.

### Fixed Issues

$ https://github.com/Expensify/Expensify/issues/198446

### Tests

Automated tests cover this change.

Manual tests:
1. This PR will have `InternalQA` label
2. Once merged and deployed to staging, it should appear in the [latest Staging QA issue](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3AStagingDeployCash)
3. Check the PR off in the list signaling that it has been resolved
4. Create another PR with a dummy change at some markdown file for example
5. Add `CP Staging` and `InternalQA` labels and merge it
6. Once it is deployed to production it should show up in the InternalQA section too and this PR should preserve its QA state (ie QA done, resolved)

- [ ] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps

Check the Tests section.

- [ ] Verify that no errors appear in the JS console

### Screenshots

N/A

### CODE REVIEW

File: .github/actions/javascript/createOrUpdateStagingDeploy/createOrUpdateStagingDeploy.js [hunk 1]
```diff
@@ -122,13 +122,20 @@ const run = function () {
                 'number',
             );
 
+            // Get the internalQA PR list, preserving the previous state of `isResolved`
+            const internalQAPRList = _.sortBy(
+                currentStagingDeployCashData.internalQAPRList,
+                'number',
+            );
```
Comment (mountiny) [inline .github/actions/javascript/createOrUpdateStagingDeploy/createOrUpdateStagingDeploy.js:129]: We are first getting the list of the InternalQA PRs from the previous issue body.

File: .github/actions/javascript/createOrUpdateStagingDeploy/createOrUpdateStagingDeploy.js [hunk 2]
```diff
@@ -122,13 +122,20 @@ const run = function () {
...
             return GithubUtils.generateStagingDeployCashBody(
                 newTag,
                 _.pluck(PRList, 'url'),
                 _.pluck(_.where(PRList, {isVerified: true}), 'url'),
                 _.pluck(_.where(PRList, {isAccessible: true}), 'url'),
                 _.pluck(deployBlockers, 'url'),
                 _.pluck(_.where(deployBlockers, {isResolved: true}), 'url'),
+                _.pluck(_.where(internalQAPRList, {isResolved: true}), 'url'),
                 didVersionChange ? false : currentStagingDeployCashData.isTimingDashboardChecked,
                 didVersionChange ? false : currentStagingDeployCashData.isFirebaseChecked,
```
Comment (mountiny) [inline .github/actions/javascript/createOrUpdateStagingDeploy/createOrUpdateStagingDeploy.js:140]: And then passing as a parameter an array of URLs of only those PRs which are resolved (the checkbox has been ticked)

File: .github/libs/GithubUtils.js [hunk 3]
```diff
@@ -273,11 +276,12 @@ class GithubUtils {
                     });
                 }
 
+                // Internal QA PR list
                 if (!_.isEmpty(internalQAPRMap)) {
                     issueBody += '\r\n\r\n\r\n**Internal QA:**';
                     _.each(internalQAPRMap, (assignees, URL) => {
                         const assigneeMentions = _.reduce(assignees, (memo, assignee) => `${memo} @${assignee}`, '');
-                        issueBody += `\r\n${_.contains(verifiedOrNoQAPRs, URL) ? '- [x]' : '- [ ]'} `;
+                        issueBody += `\r\n${_.contains(resolvedInternalQAPRs, URL) ? '- [x]' : '- [ ]'} `;
```
Comment (mountiny) [inline .github/libs/GithubUtils.js:284]: And finally check if the URL is in the `resolvedInternalQAPRs` array as the `verifiedOrNoQAPRs` does not include the necessary InternalQA PRs with `isResolved` field, hence it was always false and it have reset the checkbox to unchecked state.

File: tests/unit/GithubUtilsTest.js [hunk 4]
```diff
@@ -533,7 +534,7 @@ describe('GithubUtils', () => {
         ));
 
         test('Test some verified internalQA PRs', () => (
-            githubUtils.generateStagingDeployCashBody(tag, [...basePRList, ...internalQAPRList], [internalQAPRList[0]])
+            githubUtils.generateStagingDeployCashBody(tag, [...basePRList, ...internalQAPRList], [], [], [], [], [internalQAPRList[0]])
```
Comment (mountiny) [inline tests/unit/GithubUtilsTest.js:537]: We need to pass this new argument containing the resolved/verified InternalQA PR to the function in the test to make sure that the checkbox state has been preserved.

Comment (mountiny) [reply tests/unit/GithubUtilsTest.js:537]: The reason this test has been passing before (although this feature clearly does not work) is that we have not been simulating the actual flow as it happens in reality. We have passed in the checked-off internalQA PR as a third argument, which then made it to the [verifiedOrNoQAPRs](https://github.com/Expensify/App/blob/27031e108f9f76486c456198b7448fbbbc3afa43/.github/libs/GithubUtils.js#L251) array which then worked, however, normally, the internalQA PR would never make it to the third argument 😢

**Comment by mountiny** [comment]
This should be ready for a review, thank you very much for having a look. I will kindly ask @roryabraham and @yuwenmemon for reviews as you have worked on this before.

File: tests/unit/GithubUtilsTest.js (hunk 4 above)
Comment (roryabraham) [reply tests/unit/GithubUtilsTest.js:537]: Okay, makes sense 👍

**Review by roryabraham — DISMISSED** [review]
This looks good. Mind if we test on the next checklist? Don't want to introduce potential breaking changes or distractions when we seem close to deploying the current (massive) checklist to production.

File: .github/actions/javascript/awaitStagingDeploys/index.js [hunk 5]
```diff
@@ -390,11 +393,13 @@ class GithubUtils {
...
                 }
 
+                // Internal QA PR list
                 if (!_.isEmpty(internalQAPRMap)) {
+                    console.log('Found the following verified Internal QA PRs:', resolvedInternalQAPRs);
                     issueBody += '\r\n\r\n\r\n**Internal QA:**';
                     _.each(internalQAPRMap, (assignees, URL) => {
                         const assigneeMentions = _.reduce(assignees, (memo, assignee) => `${memo} @${assignee}`, '');
-                        issueBody += `\r\n${_.contains(verifiedOrNoQAPRs, URL) ? '- [x]' : '- [ ]'} `;
+                        issueBody += `\r\n${_.contains(resolvedInternalQAPRs, URL) ? '- [x]' : '- [ ]'} `;
```
Comment (neil-marcellini) [inline .github/actions/javascript/awaitStagingDeploys/index.js:402]: Is it possible that one of these internalQA PRs is verified but not included in `resolvedInternalQAPRs`?

File: .github/actions/javascript/createOrUpdateStagingDeploy/index.js [hunk 6]
```diff
@@ -582,11 +592,13 @@ class GithubUtils {
...
                 }
 
+                // Internal QA PR list
                 if (!_.isEmpty(internalQAPRMap)) {
+                    console.log('Found the following verified Internal QA PRs:', resolvedInternalQAPRs);
                     issueBody += '\r\n\r\n\r\n**Internal QA:**';
                     _.each(internalQAPRMap, (assignees, URL) => {
                         const assigneeMentions = _.reduce(assignees, (memo, assignee) => `${memo} @${assignee}`, '');
-                        issueBody += `\r\n${_.contains(verifiedOrNoQAPRs, URL) ? '- [x]' : '- [ ]'} `;
+                        issueBody += `\r\n${_.contains(resolvedInternalQAPRs, URL) ? '- [x]' : '- [ ]'} `;
```
Comment (neil-marcellini) [inline .github/actions/javascript/createOrUpdateStagingDeploy/index.js:601]: Why is there so much code duplication for creating the internal QA PR list and `getStagingDeployCashData` in these Github actions? I'm very new to Github actions.

**Review by neil-marcellini — COMMENTED** [review]
Due to my current limited knowledge about GH actions, I asked a few questions and I'll leave it to others to approve.

File: .github/actions/javascript/createOrUpdateStagingDeploy/index.js (hunk 6 above)
Comment (roryabraham) [reply .github/actions/javascript/createOrUpdateStagingDeploy/index.js:601]: Check out the documentation [here](https://github.com/Expensify/App/blob/main/.github/workflows/README.md#actions)

File: .github/actions/javascript/checkDeployBlockers/index.js [hunk 7]
```diff
@@ -305,6 +307,7 @@ class GithubUtils {
         accessiblePRList = [],
         deployBlockers = [],
         resolvedDeployBlockers = [],
+        resolvedInternalQAPRs = [],
```
[the remark this answers is not in the document]
Comment (mountiny) [reply .github/actions/javascript/checkDeployBlockers/index.js:310]: I agree and I must admit that when I was trying to look into this it took me a while to orient myself in the code, probably breaking it up into different parameters would be good. I would leave this for another clean up once we will need to make changes to this code to avoid breaking this even more haha

File: .github/actions/javascript/awaitStagingDeploys/index.js (hunk 5 above)
Comment (mountiny) [reply .github/actions/javascript/awaitStagingDeploys/index.js:402]: Here I guess we could have used also word `verified`,  it is just that for deploy blockers we use resolved and in the end, the meaning is the same.

We will get the `resolvedInternalQAPRs` parameter from here:
https://github.com/Expensify/App/blob/b82ca7cc5d2621654370789c0618eec34d463daf/.github/actions/javascript/createOrUpdateStagingDeploy/createOrUpdateStagingDeploy.js#L138

That comes from this call:
https://github.com/Expensify/App/blob/b82ca7cc5d2621654370789c0618eec34d463daf/.github/actions/javascript/createOrUpdateStagingDeploy/createOrUpdateStagingDeploy.js#L125-L129

Which comes from here:
https://github.com/Expensify/App/blob/0997c5fc75919e7043c002133c4b0bd9307944f4/.github/libs/GithubUtils.js#L112

And that is ultimately parsed from the issue body in this method:
https://github.com/Expensify/App/blob/0997c5fc75919e7043c002133c4b0bd9307944f4/.github/libs/GithubUtils.js#L176-L200

This is what I meant it is hard to follow 😄 The internalQA PR should not be part of the `verifiedOrNoQAPRs`but if it is resolved/verified (ie the checkbox was ticked), it will be part of the `resolvedInternalQAPRs`.

**Comment by mountiny** [comment]
Yes, let's test it. I have updated the checklist just now to remove the one internalQA PR which was there from the PR list (which is a bug and it could mess with this too). I will merge now and see how things go.

**Comment by mountiny** [comment]
@roryabraham  @neil-marcellini  Updated this PR with a fix to the additional bug, which I thought for some reason was gone haha. Left comments explaining the changes and why it was failing.

File: .github/libs/GithubUtils.js [hunk 8]
```diff
@@ -143,8 +144,7 @@ class GithubUtils {
                 isAccessible: match[4] === 'x',
             }),
         );
-        const internalQAPRList = this.getStagingDeployCashInternalQA(issue);
-        return _.sortBy(_.union(PRList, internalQAPRList), 'number');
+        return _.sortBy(PRList, 'number');
```
Comment (mountiny) [inline .github/libs/GithubUtils.js:147]: Actually, this is what was causing the internalQA PRs to be added to the normal PR list as well as the internal QA section. It is not necessary to be added here since the PRs are already fetched and then filtered somewhere else. This should fix that bug.

**Comment by mountiny** [comment]
And yes, we can test it, but it on this checklist!

**Review by roryabraham — APPROVED** [review]
This LGTM, but also I feel like maybe this code has hit a critical mass of complexity where we need to do some refactoring. If you want we can do the refactoring separately. I'll leave that up to you.

File: .github/actions/javascript/checkDeployBlockers/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (mountiny) [reply .github/actions/javascript/checkDeployBlockers/index.js:309]: @roryabraham I am not against doing it sooner than later, but I would say in a new issue and PR so this internalQA bug is fixed. 

I will be happy to work on the refactoring, but got some other stuff lined up right so I am not sure how soon I could get that coverd and we dont want this bug to annoy deployer for much longer 😄

[the remark this answers is not in the document]
Comment (mountiny) [reply .github/actions/javascript/checkDeployBlockers/index.js:309]: @roryabraham Created the issue here https://github.com/Expensify/App/issues/9999

[the remark this answers is not in the document]
Comment (roryabraham) [reply .github/actions/javascript/checkDeployBlockers/index.js:309]: Thanks for creating the issue. Pretty cool issue number #9999 too 🙂

**Comment by mountiny** [comment]
Going to merge and test this using a Test deploy blocker! 

2 bugs to test, they have the same testing steps:

1. Merge this PR and wait a bit to make sure the workflows from this will be used
2. Make sure the current checklist does not contain the internalQA amongst the _normal_ PRs (which is a big now)
3. If it does, manually remove it
4. Check off the internalQA PR which is currently listed in the internal QA section if it is not already (remember to put it back to the original state once testing it done)
5. Create a test issue and assign a DeployBlocker label to it
6. This will trigger the workflow to update the checklist with the new deploy blocker

### InternalQA is present in the list of other PRs too
- We already have one internalQA PR in the current locked checklist
- Currently, when the checklist is updated, the internalQA PR will be added to the list of _normal_ PRs which is a bug
- Make sure that the internalQA PR is **not** added to the list of normal PRs

### Preserving the checked state of the internalQA PR
- The internalQA Pr should be checked off as verified before a new deploy blocker is created
- Once the checklist is updated with the new deploy blocker, make sure the PR is still marked as verified

That is it 🎉  happy testing!

**Comment by mountiny** [comment]
Given we will QA it now, internalQA is not correct for this PR, removing and will report here how the testing went!

**Comment by mountiny** [comment]
Not an emergency. Tests were passing.

**Comment by mountiny** [comment]
First part worked: InternalQA is present in the list of other PRs too

Second part did not 😡 Preserving the checked state of the internalQA PR

Looking into the logs

**Comment by mountiny** [comment]
Oh, found the problem, from the logs:
```
Filtering out the following automated pull requests: [ 'https://github.com/Expensify/App/pull/9997' ]
Found the following Internal QA PRs: {
  'https://github.com/Expensify/App/pull/9641': [ 'PauloGasparSv', 'kidroca' ]
}
Found the following NO QA PRs: [
  'https://github.com/Expensify/App/pull/9913',
  'https://github.com/Expensify/App/pull/9890'
]
Found the following verified Internal QA PRs: [
  'https://github.com/Expensify/App/pull/9641 - @PauloGasparSv @kidroca'
]
Successfully updated StagingDeployCash! 🎉 https://github.com/Expensify/App/issues/9970
```

You can see the `Found the following verified Internal QA PRs` are not just URLs, which is wrong. I am putting up a PR to fix it!

**Comment by mountiny** [comment]
The follow up PR fixed it! https://github.com/Expensify/App/pull/10007 🎉
